codersdk/richparameters.go

```diff
@@ -116,3 +116,63 @@ func validationEnabled(param TemplateVersionParameter) bool {
 		param.Type == "bool" || // boolean type doesn't have any custom validation rules, but the value must be checked (true/false).
 		param.Type == "list(string)" // list(string) type doesn't have special validation, but we need to check if this is a correct list.
 }
+
+// ParameterResolver should be populated with legacy workload and rich parameter values from the previous build.  It then
+// supports queries against a current TemplateVersionParameter to determine whether a new value is required, or a value
+// correctly validates.
+type ParameterResolver struct {
+	Legacy []Parameter
+	Rich   []WorkspaceBuildParameter
+}
+
+// ValidateResolve checks the provided value, v, against the parameter, p, and the previous build.  If v is nil, it also
+// resolves the correct value.  It returns the value of the parameter, if valid, and an error if invalid.
+func (r *ParameterResolver) ValidateResolve(p TemplateVersionParameter, v *WorkspaceBuildParameter) (value string, err error) {
```

Comment:
I know that we're testing this logic mostly in the coderd API tests, but maybe a few extra unit tests here will be useful.